src/PerfView/RuntimeLoaderStats.cs

```diff
@@ -0,0 +1,219 @@
+                            var eventData = perThreadData[i];
+                            double startTime = eventData.Start.Time;
+                            double endTime = eventData.End.Time;
+                            double inclusiveTime = endTime - startTime;
+                            double exclusiveTime = inclusiveTime;
+                            string inclusiveTimeStr = inclusiveTime.ToString("F3");
+
+                            if (perThreadData.Count > (i + 1))
+                            {
+                                double startOfNextItem = perThreadData[i + 1].Start.Time;
+                                if (startOfNextItem < endTime)
+                                {
+                                    exclusiveTime = startOfNextItem - startTime;
+                                }
+                            }
+
+                            if (seenEvents.Contains(eventData.End.EventId))
+                                inclusiveTimeStr = "";
+
+                            writer.Write($"{PadIfNotCsv(threadId.ToString(), 12)},{PadIfNotCsv(startTime.ToString("F3"), 12)},{PadIfNotCsv(inclusiveTimeStr, 11)},{PadIfNotCsv(exclusiveTime.ToString("F3"), 11)},");
+
+                            StringBuilder eventName = new StringBuilder();
+
+                            int stackDepth = eventData.StackDepth;
+                            for (int iStackDepth = 0; iStackDepth < stackDepth; iStackDepth++)
+                                eventName.Append(" |");
+
+                            if (seenEvents.Contains(eventData.End.EventId))
+                                eventName.Append(" +");
+                            else
+                                eventName.Append("--");
+
+                            eventName.Append(eventData.Name);
+                            writer.WriteLine($"{QuoteIfCsv(eventName.ToString())},{QuoteIfCsv(process.Name)}");
+                            seenEvents.Add(eventData.End.EventId);
+                        }
+                    }
+                }
+            }
+
+            return;
+
+            string PadIfNotCsv(string str, int pad)
+            {
+                if (!csv)
+                {
+                    return str.PadLeft(pad);
+                }
+                return str;
+            }
+
+            string QuoteIfCsv(string str)
+            {
+                if (csv)
+                {
+                    return $"\"'{str}\"";
+                }
+                else
+                {
+                    return str;
+                }
+            }
+        }
+
+        public static double TotalCPUMSec(Microsoft.Diagnostics.Tracing.Analysis.TraceProcess incompleteStatsProc, Microsoft.Diagnostics.Tracing.RuntimeLoaderStats runtimeOps)
+        {
+            TraceProcess process = null;
+
+            foreach (var proc in runtimeOps.EventSource.TraceLog.Processes)
+            {
+                if (proc.ProcessID == incompleteStatsProc.ProcessID)
```

Comment:
There are a couple of ways that you can match processes, while handing the fact that process IDs can be re-used.  One is to use TraceProcess.ProcessIndex, which is a unique value across the entire trace.  The second is to use ProcessID and timestamp, and find the closest match - TraceLog.Processes has an API for this.  If you leave this as is, it's possible that PID re-use will impact the results here.

Reply:
In RuntimeLoaderStats.cs, you are using Microsoft.Diagnostics.Tracing.Etlx.TraceProcess, but here you're using Microsoft.Diagnostics.Tracing.Analysis.TraceProcess.  They are slightly different.  The first is layered on top of TraceLog, while this one is a level lower at TraceEventSource.  Since you already require Etlx.TraceProcess, I'd recommend just using that everywhere, and then you can use ProcessIndex rather than PID.

---
In reply to: [458422874](https://github.com/microsoft/perfview/pull/1232#discussion_r458422874) [](ancestors = 458422874)